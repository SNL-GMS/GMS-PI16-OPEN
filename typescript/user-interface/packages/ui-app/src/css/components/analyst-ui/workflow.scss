@import '../../gms-colors.scss';
@import '../../gms-fonts.scss';

$BLOCK_HEIGHT_PX: 24px;
$ROW_COMPUTED_HEIGHT_PX: 26px;
$BLOCK_WIDTH_PX: 140px;
$TABLE_MARGIN: 34px;
$DIVIDER_MARGIN: 31px;
//The table margin plus some manual fiddling
$WIDTH_OF_LEFT_CURTAIN: 43px;
$gap-between-rows: 0.75em;

.gms-workflow-time-axis {
  path,
  line {
    fill: none;
    stroke: $gms-prominent;
    shape-rendering: crispEdges;
  }
  text {
    font-family: $gms-sans;
    font-size: 12px;
    fill: $gms-main;
    &.day-label {
      stroke: $workflow-day-divider;
      font-size: 13px;
    }
  }
}

.workflow-display-window {
  height: 100%;
}

.workflow-panel {
  height: 100%;
  width: 100%;
}

.workflow-scroll-wrapper {
  max-height: calc(100% - 55px);
  overflow-y: scroll;
}

.interval-table-wrapper {
  flex: 1 0 auto;
  display: flex;
  margin-left: $TABLE_MARGIN;
  margin-right: $BLOCK_WIDTH_PX;
  flex-direction: column;
  overflow-x: scroll;
  overflow-y: hidden;
  position: static;
  gap: 2px;
}

.interval-table-wrapper-wrapper {
  position: relative;
  overflow-y: hidden;
  overflow-x: hidden;
}

// Left is set to account for the workflow-table-wrapper's left margin, minus the size of the button
.stage-row__expand-button {
  height: $BLOCK_HEIGHT_PX;
  border: 1px solid $gms-recessed;
  width: $BLOCK_HEIGHT_PX;
  transition: margin-bottom 0.15s;
  min-height: $ROW_COMPUTED_HEIGHT_PX;
  min-width: $ROW_COMPUTED_HEIGHT_PX;
  padding: 0;
  position: absolute;
  left: 7px;
  margin-top: -1px;
  z-index: 5;
}

.stage-row {
  display: flex;
  gap: 2px;
}

.stage-column {
  display: flex;
  flex-direction: column;
  gap: 2px;
}

.internal-row {
  display: grid;
  grid-auto-columns: 8px;
  gap: 2px;
}

.workflow-tooltip-container {
  display: grid;
  grid-template-columns: 2fr 3fr;
  margin: 0 auto;
  gap: 0.25em;
  padding-bottom: 0.5em;
  border-bottom: 1px dashed $gms-soft;
}

.workflow-tooltip-container .label-value__label {
  font-weight: bold;
}

.workflow-tooltip-container:last-child {
  border-bottom: none;
  padding-bottom: 0;
}

.workflow-tooltip-container + .workflow-tooltip-container {
  padding-top: 0.5em;
}

// The absolute position lets it float to the left over the workflow display
// The margin right and top are for alignment
.workflow-table-label {
  height: $ROW_COMPUTED_HEIGHT_PX;
  width: $BLOCK_WIDTH_PX;
  border: 1px solid;
  display: flex;
  border-color: $gms-background;
  justify-content: center;
  align-items: center;
  background: $gms-soft;
  flex: 0 0 auto;
  position: absolute;
  right: 0px;
  margin-right: 4px;
  margin-top: -1px;
  z-index: 5;
}

.workflow-table-label--activity {
  filter: $interval-activity-darken-cell-by;
}

.workflow-row {
  position: relative;
  display: flex;
  margin-bottom: 2px;
}

.interval-cell {
  display: flex;
  flex-direction: column;
  height: 100%;
  height: $BLOCK_HEIGHT_PX;
  border: 1px solid $gms-background;
  align-items: center;
  justify-content: center;
  margin: 0;
  gap: 2px;
  overflow: hidden;
  white-space: nowrap;
  cursor: pointer;
  .percent-bar {
    height: 100%;
    width: 100%;
  }
}

.interval-cell:hover {
  filter: brightness(1.2);
}

.interval-cell--in-progress {
  background-color: $interval-in-progress-fill;
  border: 1px solid $interval-in-progress-border;
  .percent-bar {
    background-color: scale-color(
      $color: $interval-in-progress-fill,
      $lightness: 25%,
      $saturation: -15%
    );
  }
}

.interval-cell--complete {
  background-color: $interval-complete-fill;
  border: 1px solid $interval-complete-border;
}

.interval-cell--not-complete {
  background-color: $interval-not-complete-fill;
  border: 1px solid $interval-not-complete-border;
}

.interval-cell--not-started {
  background-color: $interval-not-started-fill;
  border: 1px solid $interval-not-started-border;
}

.interval-cell--skipped {
  background-color: $interval-skipped-fill;
  border: 1px solid $interval-skipped-border;
}

.interval-cell--failed {
  background-color: $interval-failed-fill;
  border: 1px solid $interval-failed-border;
}

.interval-cell--selected {
  border: 2px solid $gms-selection;
}

.interval-cell--activity-cell {
  filter: $interval-activity-darken-cell-by;
}

.interval-cell--activity-cell:hover {
  filter: brightness(1);
}

.interval-cell--stale {
  filter: brightness(0.5);
}

.workflow-day-divider {
  position: absolute;
  top: 0px;
  bottom: 0px;
  width: 4px;
  z-index: 1;
  margin-left: $DIVIDER_MARGIN;
}

// To account for the offset of the workflow-table, the time axis needs to be moved to left by 1
.time-axis-wrapper {
  margin-left: -1px;
  width: calc(100%-131px);
  height: 25px;
  flex: 0 0 auto;
  position: static;
  z-index: 1;
}

.time-axis {
  width: 100%;
  height: 25px;
  overflow: hidden;
}

.interval-table__curtain-left {
  width: $WIDTH_OF_LEFT_CURTAIN;
  position: absolute;
  background-color: $gms-background;
  height: 100vh;
  z-index: 2;
  left: -6px;
}

.workflow-ellipsis {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 100px;
  text-align: center;
}

.workflow-open-anything-row {
  display: flex;
  flex-flow: wrap;
  box-sizing: border-box;
  align-items: center;
  min-width: 56ch;
  & > * {
    margin-left: 0.5rem;
  }
}

.open-anything-contents .dropdown-container {
  display: flex;
  justify-content: space-between;
}

.workflow-open-anything-row .dropdown-container .span {
  display: flex;
}

.open-anything-contents .date-range-picker {
  width: 100%;
  display: flex;
  margin-bottom: 5px;
}

.open-anything-contents .date-range-picker .date-range-picker__range-input {
  width: 65%;
  margin: 0px;
}

.open-anything-contents
  .date-range-picker
  .date-range-picker__range-input
  .bp3-popover-target
  .bp3-control-group
  * {
  flex-shrink: 1;
}

.open-anything-contents .date-range-picker .date-range-picker__label {
  font-size: 1.125em;
  font-weight: 600;
  margin: 0px;
  width: 35%;
}

.open-anything-contents .dropdown-label {
  font-size: 1.125em;
  font-weight: 600;
  margin: 0px;
  width: 35%;
}

.open-anything-contents .dropdown-selector {
  margin-bottom: 5px;
  width: 65%;
}

.interval-confirmation-contents {
  display: flex;
}

.interval-confirmation-text {
  width: 90%;
}

.interval-confirmation-icon {
  width: 10%;
}

.interval-confirmation-header {
  font-weight: bold;
}

.open-anything-error {
  color: $gms-strong-warning;
  display: flex;
  min-height: 20px;
}
.open-anything-error-text {
  margin-left: 5px;
  width: calc(100%-20px);
}
