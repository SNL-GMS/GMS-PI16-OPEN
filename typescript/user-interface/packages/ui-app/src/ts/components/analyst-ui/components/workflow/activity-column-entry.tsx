import React from 'react';

import { ActivityIntervalCell } from './activity-interval-cell';
import { ActivityColumnEntryProps } from './types';

/**
 * Row entry wrapper for ActivityInterval cells
 */
// eslint-disable-next-line react/display-name
export const ActivityColumnEntry: React.FunctionComponent<ActivityColumnEntryProps> = React.memo(
  (props: ActivityColumnEntryProps) => {
    const { activityInterval } = props;
    return <ActivityIntervalCell activityInterval={activityInterval} key={activityInterval.name} />;
  }
);
