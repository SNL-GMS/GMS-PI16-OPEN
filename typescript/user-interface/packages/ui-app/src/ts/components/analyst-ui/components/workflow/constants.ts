import { SECONDS_IN_HOUR } from '@gms/common-util';

export const PIXELS_PER_HOUR = 120;
export const PIXELS_PER_SECOND = PIXELS_PER_HOUR / SECONDS_IN_HOUR;
