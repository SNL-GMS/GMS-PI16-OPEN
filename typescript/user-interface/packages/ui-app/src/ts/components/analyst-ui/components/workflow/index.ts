export { WorkflowComponent as Workflow } from './workflow-component';
