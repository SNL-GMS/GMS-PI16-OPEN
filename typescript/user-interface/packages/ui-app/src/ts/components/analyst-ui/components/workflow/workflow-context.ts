import { WorkflowTypes } from '@gms/common-model';
import * as React from 'react';

import { OpenAnythingInterval } from './types';

export interface WorkflowContextData {
  staleStartTime: number;
  allActivitiesOpenForSelectedInterval: boolean;
  closeConfirmationPrompt: (
    interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
  ) => void;
  openConfirmationPrompt: (
    interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
  ) => void;
  openAnythingConfirmationPrompt: (interval: OpenAnythingInterval) => void;
}

/**
 * The audible notification context
 */
export const WorkflowContext: React.Context<WorkflowContextData> = React.createContext<
  WorkflowContextData
>(undefined);
