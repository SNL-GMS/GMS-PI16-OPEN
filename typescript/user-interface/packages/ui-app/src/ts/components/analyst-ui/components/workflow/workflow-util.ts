import { CommonTypes, WorkflowTypes } from '@gms/common-model';
import { isInteractiveAnalysisStage } from '@gms/common-model/lib/workflow/types';
import { epochSecondsNow } from '@gms/common-util';
import { AppState, WorkflowOperations } from '@gms/ui-state';
import { WorkflowState } from '@gms/ui-state/lib/state/analyst-workspace/workflow/types';
import Immutable from 'immutable';
import { Dispatch } from 'react';
import { MutateFunction } from 'react-query';
import { useDispatch, useSelector } from 'react-redux';

import { IanMutations, IanQueryAndMutationTypes } from '~analyst-ui/client-interface';
import { workflowQueryConfig } from '~analyst-ui/client-interface/axios/queries/workflow-query';
import { queryCache } from '~components/client-interface';

/**
 * Determines if a `stage interval` should be rendered as a percent bar.
 *
 * Auto Network is only automatic stage mode where the cells should not be percent bars
 * Auto Network time chunks are 5min so if stage mode is automatic and duration is more than
 * 5min it's not a Auto Network cell and should be shown as a percent bar
 *
 * @param interval stage interval
 * @returns boolean
 */
export const isStageIntervalPercentBar = (interval: WorkflowTypes.StageInterval): boolean => {
  const fiveMinInSeconds = 300;
  return (
    interval.endTime - interval.startTime > fiveMinInSeconds &&
    interval.stageMode === WorkflowTypes.StageMode.AUTOMATIC
  );
};

/**
 * Closes a stage
 *
 * @param userName current username
 * @param startTimeSecs open time range start time secs
 * @param openIntervalName open interval name
 * @param analystStageMutation mutation to close the stage
 */
export const closeStage = async (
  userName: string,
  startTimeSecs: number,
  openIntervalName: string,
  analystStageMutation: MutateFunction<
    void,
    unknown,
    IanQueryAndMutationTypes.UpdateInteractiveAnalysisStageIntervalStatusRequest,
    unknown
  >
): Promise<void> => {
  const args: IanQueryAndMutationTypes.UpdateInteractiveAnalysisStageIntervalStatusRequest = {
    stageIntervalId: {
      startTime: startTimeSecs,
      definitionId: {
        name: openIntervalName
      }
    },
    status: WorkflowTypes.IntervalStatus.NOT_COMPLETE,
    userName,
    time: epochSecondsNow()
  };
  await analystStageMutation(args);
};

/**
 * Determine css grid span amount by determining how scale the duration is based on
 * dividing by 5 since 5min is the smallest duration we can get
 *
 * @param startTime startTime in seconds
 * @param endTime endTime in seconds
 * @returns class name as a string to determine size
 */
export const determineCellSize = (startTime: number, endTime: number): number => {
  const sixty = 60;
  return Math.round((endTime - startTime) / sixty / 5);
};

export const setIntervalStatus = (
  dispatch: Dispatch<unknown>,
  workflow: WorkflowTypes.Workflow,
  userName: string,
  status: WorkflowTypes.IntervalStatus,
  workflowState: WorkflowState,
  activityMutation: MutateFunction<
    void,
    unknown,
    IanQueryAndMutationTypes.UpdateActivityIntervalStatusRequest,
    unknown
  >,
  analystStageMutation: MutateFunction<
    void,
    unknown,
    IanQueryAndMutationTypes.UpdateInteractiveAnalysisStageIntervalStatusRequest,
    unknown
  >
  // eslint-disable-next-line complexity
) => async (
  interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
): Promise<void> => {
  let interactiveAnalysisStage: WorkflowTypes.InteractiveAnalysisStage;
  let activities: WorkflowTypes.Activity[];
  let stageName = '';
  let isActivityInOpenStageInterval = true;

  if (WorkflowTypes.isActivityInterval(interval)) {
    stageName = interval.stageName;
  } else if (WorkflowTypes.isStageInterval(interval)) {
    stageName = interval.name;
  }

  // Close the open stage and/or activities since user is discarding changes and opening another interval
  if (
    status === WorkflowTypes.IntervalStatus.IN_PROGRESS &&
    workflowState.openIntervalName &&
    (workflowState.openIntervalName !== stageName ||
      interval.startTime !== workflowState.timeRange.startTimeSecs)
  ) {
    isActivityInOpenStageInterval = false;
    await closeStage(
      userName,
      workflowState.timeRange.startTimeSecs,
      workflowState.openIntervalName,
      analystStageMutation
    );
  }

  const foundStage: WorkflowTypes.Stage = workflow.stages.find(stage => stage.name === stageName);

  if (isInteractiveAnalysisStage(foundStage)) {
    interactiveAnalysisStage = foundStage;
  }

  if (WorkflowTypes.isStageInterval(interval)) {
    const args: IanQueryAndMutationTypes.UpdateInteractiveAnalysisStageIntervalStatusRequest = {
      stageIntervalId: {
        startTime: interval.startTime,
        definitionId: {
          name: interval.name
        }
      },
      status,
      userName,
      time: epochSecondsNow() // what should this time be?
    };
    await analystStageMutation(args);
    activities = interactiveAnalysisStage.activities;
  }

  if (WorkflowTypes.isActivityInterval(interval)) {
    const args: IanQueryAndMutationTypes.UpdateActivityIntervalStatusRequest = {
      activityIntervalId: {
        startTime: interval.startTime,
        definitionId: {
          name: interval.name
        }
      },
      stageIntervalId: {
        startTime: interval.startTime,
        definitionId: {
          name: interval.stageName
        }
      },
      status,
      userName,
      time: epochSecondsNow() // what should this time be?
    };
    await activityMutation(args);
    activities = [interactiveAnalysisStage.activities.find(act => act.name === interval.name)];
  }

  // Get the station group from the Workflow Stage
  // As guidance the first activity is the open that is supposed to be used
  const { stationGroup, analysisMode } = activities[0];
  // With a success mutation dispatch to Redux
  switch (status) {
    case WorkflowTypes.IntervalStatus.IN_PROGRESS:
      dispatch(
        WorkflowOperations.setOpenInterval(
          {
            startTimeSecs: interval.startTime,
            endTimeSecs: interval.endTime
          },
          stationGroup,
          interactiveAnalysisStage.name,
          isActivityInOpenStageInterval
            ? [...activities.map(activity => activity.name), ...workflowState.openActivityNames]
            : activities.map(activity => activity.name),
          analysisMode
        )
      );
      break;
    case WorkflowTypes.IntervalStatus.NOT_COMPLETE:
      dispatch(
        WorkflowOperations.setClosedInterval(
          activities[0].name,
          WorkflowTypes.isStageInterval(interval)
        )
      );
      break;
    // TODO add cases as they become available
    default:
  }
};

export const useSetOpenInterval = (): ((
  interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
) => Promise<void>) => {
  const dispatch = useDispatch();
  const workflow = queryCache.getQueryData<WorkflowTypes.Workflow>(workflowQueryConfig.queryKey);
  const userName = useSelector(
    (state: AppState) => state.userSessionState.authorizationStatus.userName
  );
  const workflowState = useSelector((state: AppState) => state.analystWorkspaceState.workflowState);
  const activityMutation = IanMutations.UpdateWorkflowActivityMutation.useActivityIntervalStatusMutation();
  const analystStageMutation = IanMutations.UpdateWorkflowAnalystStageMutation.useStageIntervalStatusMutation();
  return setIntervalStatus(
    dispatch,
    workflow,
    userName,
    WorkflowTypes.IntervalStatus.IN_PROGRESS,
    workflowState,
    activityMutation,
    analystStageMutation
  );
};

/**
 * Closes an interval and updates the redux state to reflect a state with no open interval
 */
export const useCloseInterval = (): ((
  interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
) => Promise<void>) => {
  const dispatch = useDispatch();
  const workflow = queryCache.getQueryData<WorkflowTypes.Workflow>(workflowQueryConfig.queryKey);
  const userName = useSelector(
    (state: AppState) => state.userSessionState.authorizationStatus.userName
  );
  const workflowState = useSelector((state: AppState) => state.analystWorkspaceState.workflowState);
  const activityMutation = IanMutations.UpdateWorkflowActivityMutation.useActivityIntervalStatusMutation();
  const analystStageMutation = IanMutations.UpdateWorkflowAnalystStageMutation.useStageIntervalStatusMutation();
  return setIntervalStatus(
    dispatch,
    workflow,
    userName,
    WorkflowTypes.IntervalStatus.NOT_COMPLETE,
    workflowState,
    activityMutation,
    analystStageMutation
  );
};

/**
 * Retrieves the TimeRange for the provided intervals; i.e. the earliest start time and the latest end time.
 *
 * @param stageIntervals the stage intervals
 * @returns the time range for the intervals
 */
export const getTimeRangeForIntervals = (
  stageIntervals: Immutable.Map<string, WorkflowTypes.StageInterval[]>
): CommonTypes.TimeRange => {
  const timeRange: CommonTypes.TimeRange = {
    startTimeSecs: Infinity,
    endTimeSecs: -Infinity
  };

  stageIntervals.forEach(s =>
    s.forEach(interval => {
      timeRange.startTimeSecs =
        interval.startTime <= timeRange.startTimeSecs
          ? interval.startTime
          : timeRange.startTimeSecs;

      timeRange.endTimeSecs =
        interval.endTime >= timeRange.endTimeSecs ? interval.endTime : timeRange.endTimeSecs;
    })
  );
  if (timeRange.startTimeSecs === Infinity || timeRange.endTimeSecs === -Infinity) {
    return undefined;
  }
  return timeRange;
};
