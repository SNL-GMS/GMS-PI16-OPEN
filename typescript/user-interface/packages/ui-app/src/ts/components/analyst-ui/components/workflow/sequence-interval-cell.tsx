import { ContextMenu } from '@blueprintjs/core';
import { WorkflowTypes } from '@gms/common-model';
import { PercentBar } from '@gms/ui-core-components';
import { AppState } from '@gms/ui-state';
import classNames from 'classnames';
import React from 'react';
import { Provider, useSelector, useStore } from 'react-redux';

import { preventDefaultEvent } from './activity-interval-cell';
import { IntervalContextMenu } from './context-menus';
import { SequenceIntervalCellProps } from './types';
import { WorkflowContext } from './workflow-context';
import { determineCellSize, isStageIntervalPercentBar } from './workflow-util';

/**
 * A sequence interval cell used to render the smaller blocks
 */
export const SequenceIntervalCell: React.FunctionComponent<SequenceIntervalCellProps> = React.memo(
  (props: SequenceIntervalCellProps) => {
    const { stageInterval } = props;
    const workflowState = useSelector(
      (state: AppState) => state.analystWorkspaceState.workflowState
    );
    const context = React.useContext(WorkflowContext);
    const isSelected =
      workflowState.openIntervalName === stageInterval.name &&
      workflowState.timeRange?.startTimeSecs === stageInterval.startTime;
    const isStale = context.staleStartTime > stageInterval.startTime;
    const cellClass = classNames({
      'interval-cell': true,
      'interval-cell--selected': isSelected,
      'interval-cell--not-complete':
        stageInterval.status === WorkflowTypes.IntervalStatus.NOT_COMPLETE,
      'interval-cell--in-progress':
        stageInterval.status === WorkflowTypes.IntervalStatus.IN_PROGRESS,
      'interval-cell--not-started':
        stageInterval.status === WorkflowTypes.IntervalStatus.NOT_STARTED,
      'interval-cell--complete': stageInterval.status === WorkflowTypes.IntervalStatus.COMPLETE,
      'interval-cell--skipped': stageInterval.status === WorkflowTypes.IntervalStatus.SKIPPED,
      'interval-cell--failed': stageInterval.status === WorkflowTypes.IntervalStatus.FAILED,
      'interval-cell--stale': isStale
    });
    const store = useStore();
    return !isStageIntervalPercentBar(stageInterval) ? (
      <div
        key={stageInterval.startTime}
        className={cellClass}
        onContextMenu={
          // We only want automatic stage intervals to show a context menu, other wise a right click should be a noop
          stageInterval.stageMode === WorkflowTypes.StageMode.INTERACTIVE && !isStale
            ? e => {
                e.preventDefault();
                ContextMenu.show(
                  <Provider store={store}>
                    <IntervalContextMenu
                      interval={stageInterval}
                      isSelectedInterval={isSelected}
                      allActivitiesOpenForSelectedInterval={
                        context.allActivitiesOpenForSelectedInterval
                      }
                      openCallback={context.openConfirmationPrompt}
                      closeCallback={context.closeConfirmationPrompt}
                    />
                  </Provider>,
                  {
                    left: e.clientX,
                    top: e.clientY
                  },
                  undefined,
                  true
                );
              }
            : preventDefaultEvent
        }
        style={{
          gridColumn: `span ${determineCellSize(stageInterval.startTime, stageInterval.endTime)}`
        }}
      />
    ) : (
      <div
        className={cellClass}
        style={{
          gridColumn: `span ${determineCellSize(stageInterval.startTime, stageInterval.endTime)}`
        }}
        onContextMenu={preventDefaultEvent}
      >
        <PercentBar percentage={stageInterval.percentAvailable} />
      </div>
    );
  }
);
