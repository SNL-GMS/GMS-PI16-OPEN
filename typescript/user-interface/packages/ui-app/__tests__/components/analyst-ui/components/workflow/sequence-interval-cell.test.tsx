import { createStore } from '@gms/ui-state';
import * as Enzyme from 'enzyme';
import * as React from 'react';
import { Provider } from 'react-redux';

import { SequenceIntervalCell } from '../../../../../src/ts/components/analyst-ui/components/workflow/sequence-interval-cell';
import { WorkflowContext } from '../../../../../src/ts/components/analyst-ui/components/workflow/workflow-context';
import { isStageIntervalPercentBar } from '../../../../../src/ts/components/analyst-ui/components/workflow/workflow-util';
import * as WorkflowDataTypes from './workflow-data-types';

const store: any = createStore();

describe('Sequence Interval Cell', () => {
  it('is exported', () => {
    expect(SequenceIntervalCell).toBeDefined();
  });

  it('matches snapshot', () => {
    const component = Enzyme.mount(
      <Provider store={store}>
        <WorkflowContext.Provider
          value={{
            staleStartTime: 1,
            allActivitiesOpenForSelectedInterval: false,
            openConfirmationPrompt: jest.fn(),
            openAnythingConfirmationPrompt: jest.fn(),
            closeConfirmationPrompt: jest.fn()
          }}
        >
          <SequenceIntervalCell
            stageInterval={WorkflowDataTypes.automaticProcessingStageInterval}
          />
        </WorkflowContext.Provider>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  it('shallow mounts', () => {
    const shallow = Enzyme.shallow(
      <Provider store={store}>
        <WorkflowContext.Provider
          value={{
            staleStartTime: 1,
            allActivitiesOpenForSelectedInterval: false,
            openConfirmationPrompt: jest.fn(),
            openAnythingConfirmationPrompt: jest.fn(),
            closeConfirmationPrompt: jest.fn()
          }}
        >
          <SequenceIntervalCell
            stageInterval={WorkflowDataTypes.automaticProcessingStageInterval}
          />
        </WorkflowContext.Provider>
      </Provider>
    );
    expect(shallow).toMatchSnapshot();
  });

  it('can determine cell percent bar', () => {
    const interactiveStageInterval = isStageIntervalPercentBar(
      WorkflowDataTypes.interactiveAnalysisStageInterval
    );
    expect(interactiveStageInterval).toBeFalsy();

    const automaticStageInterval = isStageIntervalPercentBar(
      WorkflowDataTypes.automaticProcessingStageInterval
    );
    expect(automaticStageInterval).toBeTruthy();
  });

  it('checking on context menu for automated interval', () => {
    const component = Enzyme.mount(
      <Provider store={store}>
        <WorkflowContext.Provider
          value={{
            staleStartTime: 1,
            allActivitiesOpenForSelectedInterval: false,
            openConfirmationPrompt: jest.fn(),
            openAnythingConfirmationPrompt: jest.fn(),
            closeConfirmationPrompt: jest.fn()
          }}
        >
          <SequenceIntervalCell
            stageInterval={WorkflowDataTypes.automaticProcessingStageInterval}
          />
        </WorkflowContext.Provider>
      </Provider>
    );

    component.setProps({ onContextMenu: jest.fn() });
    const seqIntItem = component.find('PercentBar');
    seqIntItem.simulate('click');
    const sequenceIntervalItem = component.find('div').first();
    const spy = jest.spyOn(sequenceIntervalItem.props(), 'onContextMenu');
    sequenceIntervalItem.simulate('click', { metaKey: true });
    expect(spy).toHaveBeenCalledTimes(0);
    const cMenu = component.find('IntervalContextMenu').first();
    expect(cMenu).toBeDefined();
    component.setProps({ stageInterval: WorkflowDataTypes.stageIntervalAuto });
    component.update();
    expect(component.props().stageInterval.stageMode).toBe(WorkflowDataTypes.automaticStageMode);
  });
});
