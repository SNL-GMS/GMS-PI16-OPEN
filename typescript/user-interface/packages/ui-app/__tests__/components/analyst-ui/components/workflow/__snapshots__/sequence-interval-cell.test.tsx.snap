// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequence Interval Cell matches snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Memo()
    stageInterval={
      Object {
        "comment": "interval example",
        "endTime": 1622057187,
        "modificationTime": 1622057667,
        "name": "Auto Network",
        "percentAvailable": 100,
        "processingEndTime": 1622057547,
        "processingStartTime": 1622057487,
        "sequenceIntervals": Array [
          Object {
            "comment": "interval example",
            "endTime": 1622057187,
            "lastExecutedStepName": "last-step",
            "modificationTime": 1622057667,
            "name": "Auto Network Seq",
            "percentAvailable": 100,
            "percentComplete": 50,
            "processingEndTime": 1622057547,
            "processingStartTime": 1622057487,
            "stageName": "Auto Network",
            "startTime": 1622053587,
            "status": "IN_PROGRESS",
            "storageTime": 1622057607,
          },
        ],
        "stageMetrics": Object {
          "associatedSignalDetectionCount": 34,
          "eventCount": 21,
          "maxMagnitude": 8,
          "unassociatedSignalDetectionCount": 55,
        },
        "stageMode": "AUTOMATIC",
        "startTime": 1622053587,
        "status": "IN_PROGRESS",
        "storageTime": 1622057607,
      }
    }
  >
    <div
      className="interval-cell interval-cell--in-progress"
      onContextMenu={[Function]}
      style={
        Object {
          "gridColumn": "span 12",
        }
      }
    >
      <PercentBar
        percentage={100}
      >
        <div
          className="percent-bar"
          key="1"
          style={
            Object {
              "transform": "translateX(-0%)",
              "width": "100%",
            }
          }
        />
      </PercentBar>
    </div>
  </Memo()>
</Provider>
`;

exports[`Sequence Interval Cell shallow mounts 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <ContextProvider
    value={
      Object {
        "allActivitiesOpenForSelectedInterval": false,
        "closeConfirmationPrompt": [MockFunction],
        "openAnythingConfirmationPrompt": [MockFunction],
        "openConfirmationPrompt": [MockFunction],
        "staleStartTime": 1,
      }
    }
  >
    <Memo()
      stageInterval={
        Object {
          "comment": "interval example",
          "endTime": 1622057187,
          "modificationTime": 1622057667,
          "name": "Auto Network",
          "percentAvailable": 100,
          "processingEndTime": 1622057547,
          "processingStartTime": 1622057487,
          "sequenceIntervals": Array [
            Object {
              "comment": "interval example",
              "endTime": 1622057187,
              "lastExecutedStepName": "last-step",
              "modificationTime": 1622057667,
              "name": "Auto Network Seq",
              "percentAvailable": 100,
              "percentComplete": 50,
              "processingEndTime": 1622057547,
              "processingStartTime": 1622057487,
              "stageName": "Auto Network",
              "startTime": 1622053587,
              "status": "IN_PROGRESS",
              "storageTime": 1622057607,
            },
          ],
          "stageMetrics": Object {
            "associatedSignalDetectionCount": 34,
            "eventCount": 21,
            "maxMagnitude": 8,
            "unassociatedSignalDetectionCount": 55,
          },
          "stageMode": "AUTOMATIC",
          "startTime": 1622053587,
          "status": "IN_PROGRESS",
          "storageTime": 1622057607,
        }
      }
    />
  </ContextProvider>
</ContextProvider>
`;
